sw-emulator/lib/cpu/src/cpu.rs

```diff
@@ -13,13 +13,53 @@ Abstract:
 --*/
 
 use crate::csr_file::{Csr, CsrFile};
+use crate::instr::Instr;
 use crate::types::{RvInstr, RvMStatus};
 use crate::xreg_file::{XReg, XRegFile};
+use bit_vec::BitVec;
 use caliptra_emu_bus::{Bus, BusError, Clock, TimerAction};
 use caliptra_emu_types::{RvAddr, RvData, RvException, RvSize};
 
 pub type InstrTracer<'a> = dyn FnMut(u32, RvInstr) + 'a;
 
+#[derive(Clone)]
+pub struct CodeCoverage {
+    bit_vec: BitVec,
+}
+
+impl CodeCoverage {
+    pub fn new(capacity_in_bytes: usize) -> Self {
+        Self {
+            bit_vec: BitVec::from_elem(capacity_in_bytes, false),
+        }
+    }
+
+    pub fn log_execution(&mut self, pc: RvData, instr: &Instr) {
+        if (pc as usize) < self.bit_vec.len() {
+            let num_bytes = match instr {
+                Instr::Compressed(_) => 2,
+                Instr::General(_) => 4,
+            };
+
+            // Mark the bytes corresponding to the executed instruction as true.
+            for i in 0..num_bytes {
+                let byte_index = (pc as usize) + i;
+                if byte_index < self.bit_vec.len() {
+                    self.bit_vec.set(byte_index, true);
+                }
+            }
+        }
+    }
+
+    pub fn count_executed(&self) -> usize {
```

Comment:
This functions aren't necessary; we'll compare to the ELF section sizes offline.

Reply:
Done.